### PR TITLE
add hd44780 driver + rppico example

### PR DESCRIPTION
**Pull request by RecursiveError**

### CODE REVIEW

File: drivers/display/hd44780.zig [hunk 1]
```diff
@@ -0,0 +1,276 @@
...
+        return Self;
+    }
+
+    pub fn reset_cursor(Self: *HD44780) *HD44780 {
+        Self.command(Lcd_commands.lcd_reset);
+        Self.internal_delay(1600); //clear and reset need to delay >1.6ms
+        return Self;
+    }
+
+    pub fn shift_cursor_left(Self: *HD44780) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:142]: see above

File: drivers/display/hd44780.zig [hunk 2]
```diff
@@ -0,0 +1,276 @@
...
+        Self.command(Lcd_commands.lcd_shift_cursor_left);
+        return Self;
+    }
+
+    pub fn shift_cursor_right(Self: *HD44780) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:147]: see above

File: drivers/display/hd44780.zig [hunk 3]
```diff
@@ -0,0 +1,276 @@
...
+        Self.command(Lcd_commands.lcd_shift_cursor_right);
+        return Self;
+    }
+
+    pub fn shift_display_left(Self: *HD44780) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:152]: see above

File: drivers/display/hd44780.zig [hunk 4]
```diff
@@ -0,0 +1,276 @@
...
+        Self.command(Lcd_commands.lcd_shift_display_left);
+        return Self;
+    }
+
+    pub fn shift_display_right(Self: *HD44780) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:157]: see above

File: drivers/display/hd44780.zig [hunk 5]
```diff
@@ -0,0 +1,276 @@
...
+        Self.command(Lcd_commands.lcd_shift_display_right);
+        return Self;
+    }
+
+    //control functions
+
+    pub fn shift_control(Self: *HD44780, st: State) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:164]: see above

File: drivers/display/hd44780.zig [hunk 6]
```diff
@@ -0,0 +1,276 @@
...
+        switch (st) {
+            .on => Self.entry_mode |= Entry_mode.lcd_shift_mode,
+            .off => Self.entry_mode &= ~Entry_mode.lcd_shift_mode,
+        }
+        _ = Self.send(Self.entry_mode, 0);
+        return Self;
+    }
+
+    pub fn shift_mode(Self: *HD44780, mode: ShiftDirection) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:173]: see above

File: drivers/display/hd44780.zig [hunk 7]
```diff
@@ -0,0 +1,276 @@
...
+        switch (mode) {
+            .inc => Self.entry_mode |= Entry_mode.lcd_shift_dir,
+            .dec => Self.entry_mode &= ~Entry_mode.lcd_shift_dir,
+        }
+        _ = Self.send(Self.entry_mode, 0);
+        return Self;
+    }
+
+    pub fn display_set(Self: *HD44780, st: State) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:182]: see above

File: drivers/display/hd44780.zig [hunk 8]
```diff
@@ -0,0 +1,276 @@
...
+        switch (st) {
+            .on => Self.display_control |= Display_control.lcd_display,
+            .off => Self.display_control &= ~Display_control.lcd_display,
+        }
+        _ = Self.send(Self.display_control, 0);
+        return Self;
+    }
+
+    pub fn cursor_control(Self: *HD44780, st: State) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:191]: see above

File: drivers/display/hd44780.zig [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+        switch (st) {
+            .on => Self.display_control |= Display_control.lcd_cursor,
+            .off => Self.display_control &= ~Display_control.lcd_cursor,
+        }
+        _ = Self.send(Self.display_control, 0);
+        return Self;
+    }
+
+    pub fn cursor_blink_control(Self: *HD44780, st: State) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:200]: see above

File: drivers/display/hd44780.zig [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+        switch (st) {
+            .on => Self.display_control |= Display_control.lcd_blink,
+            .off => Self.display_control &= ~Display_control.lcd_blink,
+        }
+        _ = Self.send(Self.display_control, 0);
+        return Self;
+    }
+
+    pub fn select_all(Self: *HD44780) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:209]: see above

File: drivers/display/hd44780.zig [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+        Self.enable_set = 0b1100;
+        return Self;
+    }
+    pub inline fn select_lcd(Self: *HD44780, en: u1) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:213]: see above

File: drivers/display/hd44780.zig [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+        Self.enable_set = 1 << en;
+        return Self;
+    }
+
+    pub fn set_cursor(Self: *HD44780, line: u8, col: u8) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:218]: see above

File: drivers/display/hd44780.zig [hunk 13]
```diff
@@ -0,0 +1,276 @@
...
+        const addrs = [_]u8{ 0x80, 0xC0 };
+        if ((line < 2) and (col < 40)) {
+            _ = Self.send(addrs[line] | col, 0);
+        }
+    }
+
+    pub fn create_custom_char(Self: *HD44780, new_char: [8]u8, mem_addr: u8) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:225]: see above

File: drivers/display/hd44780.zig [hunk 14]
```diff
@@ -0,0 +1,276 @@
...
+        const mem_aux = ((mem_addr & 0b111) << 3) | 0x40;
+        _ = Self.send(mem_aux, 0);
+        for (new_char) |line| {
+            _ = Self.send(line, 1);
+        }
+        return Self;
+    }
+
+    pub fn write(Self: *HD44780, text: []const u8) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:234]: see above

File: drivers/display/hd44780.zig [hunk 15]
```diff
@@ -0,0 +1,276 @@
...
+        for (text) |char| {
+            _ = Self.send(char, 1);
+        }
+        return Self;
+    }
+
+    pub fn init_device(Self: *HD44780, skip_init_time: bool) *HD44780 {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:241]: see above

**Review by mattnite — CHANGES_REQUESTED** [review]
Looks good, just needs style fixes

File: drivers/display/hd44780.zig [hunk 16]
```diff
@@ -0,0 +1,276 @@
...
+    //internal config vars
+    function_set: u8,
+    entry_mode: u8,
+    display_control: u8,
+    enable_set: u8,
+    full_bus: bool,
+
+    //internal_callbacks
+    internal_delay: *const lcd_delayus_callback,
+    io_interface: mdf.base.Datagram_Device,
```
Comment (ikskuh) [inline drivers/display/hd44780.zig:59]: This struct should follow the others, and have a `HD44780_Config` type which configures the `HD44780` struct. As you're using `Datagram_Device`, you should do something like
```zig
const HD44780_Config = struct {
  Datagram_Device: type = mdf.base.Datagram_Device,
};

pub fn HD44780(comptime config: HD44780_Config) type {
    return struct {
         …
         io_interface: config.Datagram_Device,
         …
    };
}
```
This way, you can either use the runtime dispatch variant with `HD44780(.{})` or you can replace the runtime dispatch with an arbitrary different dispatcher like a TestDevice or a concrete, non-vtable implementation

File: examples/raspberrypi/rp2xxx/src/rp2040_only/hd44780.zig [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+        .readv_fn = null,
+    };
+
+    pub fn datagram_device(device: *i2c_device) io_device {
+        return io_device{
+            .object = device,
+            .vtable = &vtable,
+        };
+    }
+};
```
Comment (ikskuh) [inline examples/raspberrypi/rp2xxx/src/rp2040_only/hd44780.zig:45]: We do have a ready-to-use implementation in https://github.com/ZigEmbeddedGroup/microzig/blob/dad43d5f9bad5bf8082dd2425c469e323f2066d7/port/raspberrypi/rp2xxx/src/hal/drivers.zig#L18-L34

**Review by ikskuh — CHANGES_REQUESTED** [review]
Overall it looks pretty good, but i'd remove the explicit type from `enum(u8)` if we don't require a u8 here. We can just use `enum { }` for enums that provide selective options

File: drivers/display/hd44780.zig [hunk 18]
```diff
@@ -0,0 +1,244 @@
+const mdf = @import("../framework.zig");
+pub const delayus_callback = fn (delay: u32) void;
+
+pub const LcdCommands = enum(u8) {
```
Comment (mattnite) [inline drivers/display/hd44780.zig:4]: `LCD_Commands`

**Review by mattnite — CHANGES_REQUESTED** [review]

**Review by mattnite — APPROVED** [review]

**Pull request merged**